### ISSUE TITLE
SyntaxWarning on invalid escape sequences in verdored third-party libraries (boto, pyasn1)

### ISSUE THREAD

**Issue by ttencate** [GoogleCloudPlatform/gsutil#1776]
Using Arch Linux, `google-cloud-cli` from AUR, Python 3.12.3, `gsutil` emits the following warnings every time it is run:

```
$ gsutil --version
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/__init__.py:1105: SyntaxWarning: invalid escape sequence '\c'
  """
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/pyami/config.py:98: SyntaxWarning: invalid escape sequence '\s'
  match = re.match("^#import[\s\t]*([^\s^\t]*)[\s\t]*$", line)
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:100: SyntaxWarning: invalid escape sequence '\*'
  """Create a SingleValueConstraint object.
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:141: SyntaxWarning: invalid escape sequence '\*'
  """Create a ContainedSubtypeConstraint object.
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:301: SyntaxWarning: invalid escape sequence '\*'
  """Create a PermittedAlphabetConstraint object.
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:457: SyntaxWarning: invalid escape sequence '\*'
  """Create a ConstraintsIntersection logic operator object.
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:501: SyntaxWarning: invalid escape sequence '\*'
  """Create a ConstraintsUnion logic operator object.
/opt/google-cloud-cli/platform/gsutil/third_party/pyasn1/pyasn1/type/namedval.py:15: SyntaxWarning: invalid escape sequence '\*'
  """Create named values object.
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/connection.py:672: SyntaxWarning: invalid escape sequence '\w'
  '(?:(?P<user>[\w\-\.]+):(?P<pass>.*)@)?'
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/connection.py:673: SyntaxWarning: invalid escape sequence '\w'
  '(?P<host>[\w\-\.]+)'
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/connection.py:674: SyntaxWarning: invalid escape sequence '\d'
  '(?::(?P<port>\d+))?'
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/connection.py:1121: SyntaxWarning: invalid escape sequence '\.'
  'specific endpoint this request was sent to\.'
/opt/google-cloud-cli/platform/gsutil/gslib/vendored/boto/boto/https_connection.py:80: SyntaxWarning: invalid escape sequence '\.'
  host_re = host.replace('.', '\.').replace('*', '[^.]*')
gsutil version: 5.27
```

This is probably happening because in Python 3.12 (2013-10-12, [release notes](https://docs.python.org/release/3.12.0/whatsnew/changelog.html#python-3-12-0)), invalid escape sequences turned from a `DeprecationWarning` into a `SyntaxWarning`.

**Comment by tomsykes** [comment]
This is happening for us when using gsutil as part of the cloud-sdk:alpine image during a cloud build step

**Comment by daichunqing** [comment]
Is there a workaround for that?

**Comment by ttencate** [comment]
Seems like this is no longer happening in gsutil 5.29.

**Comment by daichunqing** [comment]
@ttencate Thank you. Confirmed with gsutil 5.30.

**Comment by muffl0n** [comment]
This does not seem to be the case for Alpine. Building the image with Dockerfile https://github.com/GoogleCloudPlatform/cloud-sdk-docker/blob/master/alpine/Dockerfile with Alpine 3.20 show this error:
```
/ # cat /etc/os-release 
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.20.2
PRETTY_NAME="Alpine Linux v3.20"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://gitlab.alpinelinux.org/alpine/aports/-/issues"
/ # gsutil version
/google-cloud-sdk/platform/gsutil/gslib/vendored/boto/boto/__init__.py:1105: SyntaxWarning: invalid escape sequence '\c'
  """
/google-cloud-sdk/platform/gsutil/gslib/vendored/boto/boto/pyami/config.py:98: SyntaxWarning: invalid escape sequence '\s'
  match = re.match("^#import[\s\t]*([^\s^\t]*)[\s\t]*$", line)
/google-cloud-sdk/platform/gsutil/third_party/pyasn1/pyasn1/type/constraint.py:100: SyntaxWarning: invalid escape sequence '\*'
...
gsutil version: 5.30
```

**Comment by pavandagagoogle** [comment]
gsutil has issues with Python 3.12 which is packaged with Alpine 3.20 so it is a compatibility issue

**Comment by muffl0n** [comment]
Thanks for the feedback! I interpreted the closed issue, along with the positive feedback from @daichunqing and @ttencate, as an indication that this should have already been fixed. The original poster describes the exact incompatibility with Python 3.12.

**Comment by ttencate** [comment]
Maybe Arch distributors patched it or suppressed the warnings? Anyway, reopening.

**Comment by maxdanilov** [comment]
It's been over a year (!) since Python 3.12 was released, and still Google hasn't tackled this, smh.

**Comment by phlax** [comment]
its a super easy fix - the problem strings need to be marked as raw - ie `r"..."` - so the escapes are not interpreted

**Comment by phlax** [comment]
PR for boto here https://github.com/gsutil-mirrors/boto/pull/24

**Comment by richardburakowski** [comment]
@ttencate - google-cloud-cli from AUR (which is built from the debian package) ships with python 3.11 and is used by gcloud/gsutil by default.  You may want to investigate why you're using the system python 3.12 instead.

For me, I wanted to use numpy with gcloud ssh and was using the system installation
```
CLOUDSDK_PYTHON=/usr/bin/python
CLOUDSDK_PYTHON_SITEPACKAGES=True
```

But since the syntax warning I've moved to using the shipped 3.11
```
CLOUDSDK_PYTHON_SITEPACKAGES=True
/opt/google-cloud-cli/platform/bundledpythonunix/bin/python -m pip install numpy
```

**Comment by ttencate** [comment]
@richardburakowski Maybe it used the system Python before, and this was changed to 3.11 shortly after 3.12 was released. This would explain why the errors disappeared for me.

**Comment by mrdomino** [comment]
This is also broken on Ubuntu 24.04 and macOS when using MacPorts.

**Comment by sameer-google** [comment]
The warnings are resolved in gsutil version 5.32 and above, which supports Python 3.12. Upgrade to this version or later to fix the issue.

**Comment by joshdempster** [comment]
If it's fixed in 5.32, why does gsutil 5.33 still raise an error if you try to run it with python 3.12?
